로그인, 프로필 화면 [STEP 2] Forest, Volga

volgaHwang opened the pull request:
@wonhee009 

안녕하세요! 라자냐
Forest와 Volga입니다. Step2 PR을 작성해서 보내드립니다.

# 고민한 점

## AutoLayout
- NSLayoutConstraint를 사용해서 ConstraintLayout을 적용했습니다.
- SubView들이 늘어나는 만큼, 코드 크기가 같이 늘어나서 커스텀 뷰에서 처리할 부분들은 따로 커스텀 뷰에서 처리하게 했습니다.

## Close, Favorite, Gift, Won Button autoLayout구현
- 처음에  Hugging 값을 변경 해서 스택뷰로 구현하려고 했는데, 의도한 대로 구현이 되지않았습니다.
- button별로 따로 constraint를 적용해서 구현했습니다.

## 1:1 대화, 전화하기, 카카오스토리 CustomView
- 이미지 별 / 레이블 별 Center Align이 맞춰져있어서, 이 부분에 대해 어떻게 동시에 만족시킬수 있을지 고민했습니다.
- 추가적으로 이미지 + 레이블을 눌러도 버튼이 동작해야 하는데 이러한 점 까지 반영해서 모두 만족시킬수 있는지에 대해 고민했습니다.
- CustomView를 만들어서 Gesture를 주는 방식으로 문제를 해결했습니다.

# 궁금한 점
- StackView를 사용해서 상단의 메뉴들을 구현하는 방법이 있을지 궁금합니다
    - close button의 hugging 값만 낮춰서 실행을 해봤는데, close button만 크기가 늘어났습니다.
    - 그래서 StackView 사용하는 것을 포기했는데 이 부분에 대해 혹시 해결책이 있을까요?
- 밑에 1:1대화, 전화, 카카오스토리 를 Custom View로 만들어서 구현했는데, Custom View말고도 Image View들 / Label들 에 대해 Align을 동일하게 만족시키면서 구현하는 방법이 있을까요?

forestjae committed:
chore: gitignore 추가

forestjae committed:
feat: 로그인 화면 구성요소 추가

- 스토리보드

volgaHwang committed:
feat: ID 5글자 입력 시, 로그인 버튼 색상 변경 구현

- idTextField에 addTarget 사용해서, textFieldDidChange를 사용하게 구현했습니다.

volgaHwang committed:
feat: text에 대해 dynamicTypes 추가 구현

- setDaynmicTypes 함수로 text들에 대해 dynamic types 속성을 적용했습니다.

volgaHwang committed:
feat: 지역화 구현 반영

- Localizable 추가해서 한국어, 영어에 대한 key-value 파일을 만들어서 반영했습니다.
- LocalizedLiteral 로 namespace를 만들어서 사용하게 했습니다.

volgaHwang committed:
refactor: LocalizedLiteral / String Extension 파일 분리

- ViewController에 포함되어있던 LocalizedLiteral과 String Extension 파일을 분리했습니다.

volgaHwang committed:
refactor: 상수 부분 namespace Constant로 변경 사항 반영

- enum Constant 사용해서 namespace로 최소 카운트 5 반영하여 작성했습니다

forestjae committed:
refactor: textFieldDidChange 메서드 파라미터 삭제

forestjae committed:
feat: UI 컴포넌트 구현

- 이미지 에셋 추가

volgaHwang committed:
refactor: textFieldDidChange 파라미터 (_ sender: UITextField) 반영

- 기존에 ViewController 내부에 존재하는 idTextField를 명시적으로 사용하는 것 보다는, 메소드 동작 주체에 대해 바로 파라미터로 사용하는 방식으로 수정했습니다.

volgaHwang committed:
feat: TouchableView 구현 사항 반영

- 전화하기, 1:1채팅, 카카오스토리 부분에 대해 ImageView와 Label을 묶는 CustomView를 구현했습니다.
- setHierarchy, setLayout를 통해 뷰 구성을 변경할수 있도록 구현했습니다.
- setImage와 setLabel을 통해 서브 뷰의 구성을 변경할 수 있게 구현했습니다.

volgaHwang committed:
feat: ProfileViewController AutoLayout 구현

- AutoLayout을 Sub Views에 적용했습니다.
- ProfileViewController에서 setLayout, setHierarchy, setAttributes를 사용해서 뷰들을 조정하게 구현했습니다.

wonhee009 commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,153 @@
+
+class ProfileViewController: UIViewController {
+    
+    let lowerMenuStackView: UIStackView = {
```
`lowerMenuStackView`보다 좀 더 명확한 이름이 있을까요?
StackView여서 뷰들의 컨테이너 역할을 한다는건 알겠는데 lower만 봐서는 안에 어떤 뷰들이 포함되어 있는지는 모를거 같습니다.

wonhee009 replied:
해당 뷰뿐 아니라 아래의 button, imageView들이 전부 외부에서 접근 가능하도록 열려 있는데요. private일 필요는 없을까요?

wonhee009 commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,153 @@
+    }
+    
+    private func setLayout() {
+        // Todo: 레이아웃 정의
```
필요한 주석인가요?

volgaHwang replied:
불필요한 주석입니다.
페어 하는 동안에 목적성을 좀 확실히 하려고 적어 놓은 건데 없애는걸 잊었습니다 ㅠㅠ

wonhee009 commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,153 @@
+    }
+    
+    private func setAttributes() {
+        view.backgroundColor = .red
```
backgroundColor가 red인게 명세서 스펙인가요?

volgaHwang replied:
헛... 이 부분은 저희가 실험을 한번 해볼려고 넣어놨던 코드인데 빼먹은거같아요

wonhee009 commented on LoginProfile/TouchableView.swift:
```diff
@@ -0,0 +1,59 @@
+
+import UIKit
+
+class TouchableView: UIView {
```
TouchableView라고 네이밍한 이유가 궁금합니다.
이름만 봤을때에는 해당 뷰 내에서 터치에 관련한 처리를 할거 같은데 관련 코드가 없는거 같아서요.

wonhee009 commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,153 @@
+    let closeButton = CloseButton()
+    let wonButton = WonButton()
+    let giftButton = GiftButton()
+    let favouriteButton = FavouriteButton()
```
```suggestion
    let favoriteButton = FavouriteButton()
```

wonhee009 reviewed (commented):
수고하셨습니다 :)

타겟 브랜치 확인 부탁드려요.
코멘트 남겨두었으니 확인 부탁드립니다.

- 상단의 메뉴 전체를 stackView로 구현하고 싶은 이유가 따로 있을까요? 현재 디자인 상으로 봤을때 x 버튼과 우측의 버튼들 사이의 간격이 있어서 x버튼을 UIView로 한 번 더 감싸서 stackView에 넣어주던지 해야 x 버튼의 터치 영역이 길어지지 않을거 같긴 합니다.
- 이미지와 라벨을 vertical stackView에 넣는 방식도 있을거고요. 오토레이아웃으로 그냥 잡아주는 방식도 있을거 같습니다.

volgaHwang committed:
refactor: 프로퍼티 접근제한자, 네이밍 변경 + 주석제거

- 외부에서 접근이 불필요한 프로퍼티들에 대해 private 으로 접근제한자를 추가했습니다.
- lowerMenuStackView -> menuStackView로 변경했습니다.
- todo 같은 불필요한 주석을 제거합니다.
